contracts/p1/RevenueTrader.sol

```diff
@@ -48,24 +71,34 @@ contract RevenueTraderP1 is TradingP1, IRevenueTrader {
     //
     // If erc20 is any other registered asset (checked):
     //   actions:
-    //     tryTrade(prepareTradeSell(toAsset(erc20), toAsset(tokenToBuy), bal))
+    //     tryTrade(kind, prepareTradeSell(toAsset(erc20), toAsset(tokenToBuy), bal))
     //     (i.e, start a trade, selling as much of our bal of erc20 as we can, to buy tokenToBuy)
-    function manageToken(IERC20 erc20) external notTradingPausedOrFrozen {
-        if (address(trades[erc20]) != address(0)) return;
-
-        uint256 bal = erc20.balanceOf(address(this));
-        if (bal == 0) return;
-
+    function manageToken(IERC20 erc20, TradeKind kind) external notTradingPausedOrFrozen {
         if (erc20 == tokenToBuy) {
-            // == Interactions then return ==
-            IERC20Upgradeable(address(erc20)).safeApprove(address(distributor), 0);
-            IERC20Upgradeable(address(erc20)).safeApprove(address(distributor), bal);
-            distributor.distribute(erc20, bal);
+            distributeTokenToBuy();
             return;
         }
 
         IAsset sell = assetRegistry.toAsset(erc20);
         IAsset buy = assetRegistry.toAsset(tokenToBuy);
+
+        // === Refresh ===
+        // do not need to refresh when caller is BackingManager.forwardRevenue()
+        if (_msgSender() != address(backingManager)) {
+            if (erc20 == IERC20(address(rToken)) || tokenToBuy == IERC20(address(rToken))) {
+                // if either token is the RToken, refresh everything
+                assetRegistry.refresh();
+                furnace.melt();
+            } else {
+                // otherwise, refresh just buy + sell
+                sell.refresh();
+                buy.refresh();
+            }
+        }
+
+        require(address(trades[erc20]) == address(0), "trade open");
```

Comment:
Any reason to why they were moved here? Ideally these would be the first checks in order to fail fast. These are reverts anyway, so previous progress is always lost.

Reply:
Agree, ideally they'd be first. This is an instance though of our larger RCEI pattern, where it's important to do the checks _after_ the first set of refresh interactions, since they could plausibly change the outcome of the checks. 

This wasn't labeled well as trying to follow RCEI, so I've added that to the `@custom:interaction` comment at the top of the function. The `tokenToBuy` is immutable after init and the assetRegistry assets can only be changed under governance, which is why it's acceptable to place these lines above the refresh() interactions.

Reply:
aaaand that should be in a comment shouldn't it

Reply:
Actually, I want to get rid of even that. No reason to settle for reetrancy-safe-unless-governance-action when we can get the real thing.